File: backend/app/Modules/Asset/Models/Asset.php
<?php

namespace App\Modules\Asset\Models;

use App\Models\Borrow;
use App\Modules\AssetCategory\Models\AssetCategory;
use App\Modules\AssetIdentifier\Models\AssetIdentifier;
use App\Modules\Asset\Enums\AssetStatus;
use App\Modules\Asset\Enums\ConditionStatus;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\SoftDeletes;

class Asset extends Model
{
    use HasFactory;
    use SoftDeletes;

    /**
     * Resolve the factory for this model.
     * The factory lives at Database\Factories\AssetFactory (legacy location).
     */
    protected static function newFactory(): \Database\Factories\AssetFactory
    {
        return \Database\Factories\AssetFactory::new();
    }

    protected $fillable = [
        'asset_number',
        'property_number',
        'name',
        'description',
        'asset_category_id',
        'manufacturer_id',
        'office_id',
        'location_id',
        'model',
        'status',
        'condition_status',
        'purchase_date',
        'purchase_cost',
        'warranty_until',
        'insurance_provider',
        'insurance_policy_number',
        'insurance_expiration_date',
        'remarks',
        'issued_to',
        'issued_to_user_id',
        'issued_by_user_id',
        'date_issued',
        // Disposal lifecycle fields
        'disposal_reason',
        'disposal_date',
        'disposal_method',
        'disposal_approval_ref',
        'disposal_approved_by',
        'disposal_cancelled_at',
        'disposal_cancel_reason',
        'custodian_id',
    ];

    protected function casts(): array
    {
        return [
            'status' => AssetStatus::class,
            'condition_status' => ConditionStatus::class,
            'purchase_date' => 'date',
            'warranty_until' => 'date',
            'insurance_expiration_date' => 'date',
            'date_issued' => 'date',
            'disposal_date' => 'date',
            'disposal_cancelled_at' => 'datetime',
            'purchase_cost' => 'decimal:2',
        ];
    }

    public function category(): BelongsTo
    {
        return $this->belongsTo(AssetCategory::class, 'asset_category_id');
    }

    public function manufacturer(): BelongsTo
    {
        return $this->belongsTo(Manufacturer::class);
    }

    public function office(): BelongsTo
    {
        return $this->belongsTo(Office::class);
    }

    public function location(): BelongsTo
    {
        return $this->belongsTo(Location::class);
    }

    public function issuedByUser(): BelongsTo
    {
        return $this->belongsTo(\App\Models\User::class, 'issued_by_user_id');
    }

    public function issuedToUser(): BelongsTo
    {
        return $this->belongsTo(\App\Models\User::class, 'issued_to_user_id');
    }

    public function issuanceHistories(): HasMany
    {
        return $this->hasMany(AssetIssuanceHistory::class, 'asset_id');
    }

    public function identifiers(): HasMany
    {
        return $this->hasMany(AssetIdentifier::class);
    }

    public function attachments(): HasMany
    {
        return $this->hasMany(AssetAttachment::class);
    }

    public function borrows(): HasMany
    {
        return $this->hasMany(\App\Modules\Borrowing\Models\Borrowing::class);
    }

    /**
     * The linked InventoryItem that owns this asset record.
     * inventory_items.asset_id → assets.id (inverse of BelongsTo on InventoryItem).
     */
    public function inventoryItem(): \Illuminate\Database\Eloquent\Relations\HasOne
    {
        return $this->hasOne(\App\Modules\Inventory\Models\InventoryItem::class, 'asset_id');
    }

    /** User who approved / initiated the disposal workflow. */
    public function disposalApprover(): BelongsTo
    {
        return $this->belongsTo(\App\Models\User::class, 'disposal_approved_by');
    }

    /**
     * Current custodian (person responsible for the physical asset instance).
     */
    public function custodian(): BelongsTo
    {
        return $this->belongsTo(\App\Models\User::class, 'custodian_id');
    }

    public function scopeAvailable(Builder $query): Builder
    {
        return $query->where('status', AssetStatus::AVAILABLE);
    }

    public function scopeSearch(Builder $query, ?string $term): Builder
    {
        if ($term === null || $term === '') {
            return $query;
        }

        $like = '%'.$term.'%';
        $operator = $query->getConnection()->getDriverName() === 'pgsql' ? 'ilike' : 'like';

        return $query->where(function (Builder $builder) use ($like, $operator) {
            $builder
                ->where('asset_number', $operator, $like)
                ->orWhere('property_number', $operator, $like)
                ->orWhere('name', $operator, $like)
                ->orWhere('model', $operator, $like)
                ->orWhereHas('identifiers', function (Builder $identifiers) use ($like, $operator) {
                    $identifiers->where('identifier_value', $operator, $like);
                });
        });
    }
}

File: mobile/lib/data/models/asset.dart
class Asset {
  final int id;
  final String assetNumber;
  final String name;
  final String? description;
  final String? model;
  final String status;
  final String? conditionStatus;
  final String? purchaseDate;
  final double? purchaseCost;
  final String? warrantyUntil;
  final AssetCategory? category;
  final AssetOffice? office;
  final AssetLocation? location;
  final List<AssetIdentifier>? identifiers;

  Asset({
    required this.id,
    required this.assetNumber,
    required this.name,
    this.description,
    this.model,
    required this.status,
    this.conditionStatus,
    this.purchaseDate,
    this.purchaseCost,
    this.warrantyUntil,
    this.category,
    this.office,
    this.location,
    this.identifiers,
  });

  static double? _parseDouble(dynamic value) {
    if (value == null) return null;
    if (value is num) return value.toDouble();
    if (value is String) return double.tryParse(value);
    return null;
  }

  factory Asset.fromJson(Map<String, dynamic> json) {
    return Asset(
      id: json['id'] as int,
      assetNumber: json['asset_number'] as String,
      name: json['name'] as String,
      description: json['description'] as String?,
      model: json['model'] as String?,
      status: json['status'] as String,
      conditionStatus: json['condition_status'] as String?,
      purchaseDate: json['purchase_date'] as String?,
      purchaseCost: _parseDouble(json['purchase_cost']),
      warrantyUntil: json['warranty_until'] as String?,
      category: json['category'] != null 
          ? AssetCategory.fromJson(json['category'] as Map<String, dynamic>) 
          : null,
      office: json['office'] != null 
          ? AssetOffice.fromJson(json['office'] as Map<String, dynamic>) 
          : null,
      location: json['location'] != null 
          ? AssetLocation.fromJson(json['location'] as Map<String, dynamic>) 
          : null,
      identifiers: (json['identifiers'] as List<dynamic>?)
          ?.map((e) => AssetIdentifier.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'asset_number': assetNumber,
      'name': name,
      'description': description,
      'model': model,
      'status': status,
      'condition_status': conditionStatus,
      'purchase_date': purchaseDate,
      'purchase_cost': purchaseCost,
      'warranty_until': warrantyUntil,
      'category': category?.toJson(),
      'office': office?.toJson(),
      'location': location?.toJson(),
      'identifiers': identifiers?.map((e) => e.toJson()).toList(),
    };
  }
}

class AssetCategory {
  final int id;
  final String name;

  AssetCategory({
    required this.id,
    required this.name,
  });

  factory AssetCategory.fromJson(Map<String, dynamic> json) {
    return AssetCategory(
      id: json['id'] as int,
      name: json['name'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
    };
  }
}

class AssetLocation {
  final int id;
  final String name;

  AssetLocation({
    required this.id,
    required this.name,
  });

  factory AssetLocation.fromJson(Map<String, dynamic> json) {
    return AssetLocation(
      id: json['id'] as int,
      name: json['name'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
    };
  }
}

class AssetOffice {
  final int id;
  final String name;

  AssetOffice({
    required this.id,
    required this.name,
  });

  factory AssetOffice.fromJson(Map<String, dynamic> json) {
    return AssetOffice(
      id: json['id'] as int,
      name: json['name'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
    };
  }
}

class AssetIdentifier {
  final int id;
  final String identifierType;
  final String identifierValue;

  AssetIdentifier({
    required this.id,
    required this.identifierType,
    required this.identifierValue,
  });

  factory AssetIdentifier.fromJson(Map<String, dynamic> json) {
    return AssetIdentifier(
      id: json['id'] as int,
      identifierType: json['identifier_type'] as String,
      identifierValue: json['identifier_value'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'identifier_type': identifierType,
      'identifier_value': identifierValue,
    };
  }
}
